File: LeetCode/1020. Number of Enclaves/solution.cpp
/**
 * 57 / 57 test cases passed.
 * Runtime: 56 ms
 * Memory Usage: 21.5 MB 
 */
class Solution {
public:
    int dx[4] = {1, -1, 0, 0};
    int dy[4] = {0, 0, 1, -1};
    int m, n; 
    void dfs(vector<vector<int>> &grid, vector<vector<bool>> &used, int x, int y) {
        if (0 > x || m <= x || 0 > y || n <= y || !grid[x][y] || used[x][y]) {
            return;
        }
        used[x][y] = true;
        for (int i = 0; i < 4; ++ i) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            dfs(grid, used, nx, ny);
        }
    }
    int numEnclaves(vector<vector<int>>& grid) {
        m = grid.size();
        n = grid[0].size();
        int enclave = 0;
        vector<vector<bool>> used(m, vector<bool>(n));
        for (int i = 0; i < m; ++ i) {
            dfs(grid, used, i, 0);
            dfs(grid, used, i, n - 1);
        }
        for (int i = 1; i < n - 1; ++ i) {
            dfs(grid, used, 0, i);
            dfs(grid, used, m - 1, i);
        }
        for (int i = 1; i < m - 1; ++ i) {
            for (int j = 1; j < n - 1; ++ j) {
                if (grid[i][j] && !used[i][j]) {
                    ++ enclave;
                }
            }
        }
        return enclave;
    }
};

File: LeetCode/1020. Number of Enclaves/solution.py
"""
57 / 57 test cases passed.
Runtime: 80 ms
Memory Usage: 17 MB
"""
class Solution:
    def numEnclaves(self, grid: List[List[int]]) -> int:
        ans, m, n = 0, len(grid), len(grid[0])
        def dfs(x, y):
            grid[x][y] = 0
            for xx, yy in [(x + 1, y), (x - 1, y), (x, y + 1), (x, y - 1)]:
                if 0 <= xx < m and 0 <= yy < n and grid[xx][yy] == 1:
                    dfs(xx, yy)
        
        for i in range(m):
            if grid[i][0]: dfs(i, 0)
            if grid[i][n - 1]: dfs(i, n - 1)
        for i in range(1, n - 1):
            if grid[0][i]: dfs(0, i)
            if grid[m - 1][i]: dfs(m - 1, i)
        for i in range(1, m - 1):
            for j in range(1, n - 1):
                if grid[i][j]:
                    ans += 1
        return ans
